packages/server/src/core/bank_rule/specialization_groups.rs

```diff
@@ -1,71 +1,197 @@
+// sg = specialization_group
+// sgs = specialization_groups
+
+fn check_courses_assignment_for_sgs(
+    sgs: &[SpecializationGroup],
+    groups_indices: &[u8],
+    course_id_to_sg_index: &HashMap<CourseId, u8>,
+) -> bool {
+    for sg_index in groups_indices {
+        // check there are enough courses in this specialization group
+        if (course_id_to_sg_index
+            .values()
+            .filter(|group| *group == sg_index)
+            .count() as u8)
+            < sgs[*sg_index as usize].courses_sum
+        {
+            // There are not enough courses in this assignment to complete sg requirement
+            return false;
+        }
+        // check if the user completed the mandatory courses in sg
+        if let Some(mandatory) = &sgs[*sg_index as usize].mandatory {
+            for courses in mandatory {
+                let mut completed_current_demand = false;
+                for (course_id, group) in course_id_to_sg_index {
+                    // check if the user completed one of courses
+                    if group == sg_index && courses.contains(course_id) {
+                        completed_current_demand = true;
+                        break;
+                    }
+                }
+                if !completed_current_demand {
+                    return false;
+                }
+            }
+        }
+    }
+    true
+}
+
+// This function is looking for a valid assignment for the courses which fulfill the sgs requirements
+// If an assignment is found it returns it, None otherwise.
+fn find_valid_assignment_for_courses(
+    sgs: &[SpecializationGroup],
+    groups_indices: &[u8],
+    optional_sgs_for_course: &HashMap<CourseId, Vec<u8>>, // list of all optional sgs for each course
+    course_id_to_sg_index: &mut HashMap<CourseId, u8>,
+    course_index: usize, // course_index-th element in optional_sgs_for_course
+) -> Option<HashMap<CourseId, u8>> {
+    if course_index >= optional_sgs_for_course.len() {
+        if check_courses_assignment_for_sgs(sgs, groups_indices, course_id_to_sg_index) {
+            return Some(course_id_to_sg_index.clone());
+        }
+        return None;
+    }
+    if let Some((course_id, optional_groups)) = optional_sgs_for_course.iter().nth(course_index) {
+        for sg_index in optional_groups {
+            course_id_to_sg_index.insert(course_id.clone(), *sg_index);
+            if let Some(valid_assignment) = find_valid_assignment_for_courses(
+                sgs,
+                groups_indices,
+                optional_sgs_for_course,
+                course_id_to_sg_index,
+                course_index + 1,
+            ) {
+                return Some(valid_assignment);
+            }
+        }
+    }
+    None
+}
+
+fn check_if_completed_groups(
+    sgs: &[SpecializationGroup],
+    groups_indices: &[u8],
+    courses: &[CourseId],
+) -> Option<HashMap<CourseId, u8>> {
+    let mut optional_sgs_for_course = HashMap::<CourseId, Vec<u8>>::new();
```

Comment:
since this function optionally returns the completed groups I think a more understandable name would "get_completed_groups" or "find_completed_groups". 
1st option is more accurate while the 2nd option is more consistent with the function above "find_valid_assignment_for_courses", which also optionally returns something. 
I am find with both choices.
also, maybe change 'groups' to 'sgs' for consistency.

Reply:
Changed the name to "get_sgs_courses_assignment"